### ISSUE TITLE
using this with https://github.com/RussellLuo/caddy-ext

### ISSUE THREAD

**Issue by gedw99** [RussellLuo/olaf#1]
The Rest API exposed does not have a open api or something to work with yet ? Is there a roadmap / plan for the steps that you have in your mind..

Would be really cool to be able to use this with https://github.com/RussellLuo/caddy-ext.   I never knew Caddy had layer 4 and could do LB.. No need to use the clouds load balancers then.

**Comment by RussellLuo** [comment]
> The Rest API exposed does not have a open api or something to work with yet ?

Yeah, as described in README, it currently only supports Read-Only Admin API for Declarative configuration. I guess DB-based implementation will be added in the future.

> I never knew Caddy had layer 4 and could do LB.. No need to use the clouds load balancers then.

Yes, [caddy-l4](https://github.com/mholt/caddy-l4) is capable of doing that for you :)

**Comment by gedw99** [comment]
I don’t get why a db would help. 

A cli that allows ops to configure one ( or many ) caddy servers would be nice. That gives you direct remote control. 

On the other hand a way to write changes , like how people do db sql migration scripts would also be cool. You could checkin the script to git and then sone flux system would apply the changes to you many caddy servers .

If you have other mechanisms in mind to achieve this type of migration patterns would be really interesting to hear them.

**Comment by RussellLuo** [comment]
> I don’t get why a db would help.

Besides [Declarative Configuration](https://docs.konghq.com/gateway/latest/reference/db-less-and-declarative-config/), Kong can also [run with a database](https://docs.konghq.com/gateway/latest/reference/configuration/#datastore-section). I think this is essential for its [Admin API](https://docs.konghq.com/gateway/latest/admin-api/).

> A cli that allows ops to configure one ( or many ) caddy servers would be nice. That gives you direct remote control.

Caddy already has its own [API](https://caddyserver.com/docs/api), which I think can be used by ops as you expected.

> On the other hand a way to write changes , like how people do db sql migration scripts would also be cool. You could checkin the script to git and then sone flux system would apply the changes to you many caddy servers .

I personally think something like a migration script is "imperative" (i.e. ["do this, then do that"](https://docs.konghq.com/gateway/latest/reference/db-less-and-declarative-config/#declarative-configuration)), which is increasingly deprecated in favor of "declarative", which I guess is also the style commonly used in [Kubernetes](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/).

**Comment by gedw99** [comment]
thanks for the links about Kong. Its help me get your point... Kong approach looks fine with me.  Is all that stuff in the docs open source though, cause you know how it is - half open sometimes :) Have you used this stuff ? I might give it a try.

Does it support the flux pattern then ? when you change a config and commit it to git, you want something else to pick up the change and merge it into your many servers.  This is the flux pattern. It used to manage k8 in a CI / CD fashion. Its golang and open: https://fluxcd.io/docs/

**Comment by gedw99** [comment]
I had another go at this, and really like the way its designed and its simplicity.

When i run OLAF the web gui at localhost:2020 shows nothing.

```
olaf git:(master) ✗ ./olaf 
2022/05/13 19:04:24 transport=HTTP addr=:2020
```


Does it also need Caddy running also ?

**Comment by RussellLuo** [comment]
> When i run OLAF the web gui at localhost:2020 shows nothing.

Yeah, `cmd/olaf` is currently just a (read-only) RESTful API, which does not come with a GUI.

As an example, in order to interact with the RESTful API, try the following endpoint instead:

```bash
$ curl -i http://localhost:2020/services
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Sun, 15 May 2022 12:48:56 GMT
Content-Length: 605

[{"name":"static","upstream":null},{"name":"production","upstream":{"backends":[{"dial":"localhost:2222","max_requests":100},{"dial":"localhost:2223","max_requests":100}],"http":{"dial_timeout":"5s"},"lb":{"policy":"round_robin","try_duration":"","interval":""},"active_hc":null,"header_up":null,"header_down":{"set":null,"add":{"Server":["Production"]},"delete":null}}},{"name":"staging","upstream":{"backends":[{"dial":"localhost:3333","max_requests":100}],"http":{"dial_timeout":"5s"},"lb":null,"active_hc":null,"header_up":null,"header_down":{"set":null,"add":{"Server":["Staging"]},"delete":null}}}]
```
